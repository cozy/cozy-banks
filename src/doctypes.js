import fromPairs from 'lodash/fromPairs'
import CozyClient, {
  QueryDefinition,
  HasManyInPlace,
  Q,
  HasMany
} from 'cozy-client'
import subYears from 'date-fns/sub_years'
import format from 'date-fns/format'
import flag from 'cozy-flags'

// eslint-disable-next-line no-unused-vars
import { Connection } from './types'

export const RECIPIENT_DOCTYPE = 'io.cozy.bank.recipients'
export const ACCOUNT_DOCTYPE = 'io.cozy.bank.accounts'
export const GROUP_DOCTYPE = 'io.cozy.bank.groups'
export const TRANSACTION_DOCTYPE = 'io.cozy.bank.operations'
export const SETTINGS_DOCTYPE = 'io.cozy.bank.settings'
export const BILLS_DOCTYPE = 'io.cozy.bills'
export const TRIGGER_DOCTYPE = 'io.cozy.triggers'
export const JOBS_DOCTYPE = 'io.cozy.jobs'
export const APP_DOCTYPE = 'io.cozy.apps'
export const KONNECTOR_DOCTYPE = 'io.cozy.konnectors'
export const COZY_ACCOUNT_DOCTYPE = 'io.cozy.accounts'
export const PERMISSION_DOCTYPE = 'io.cozy.permissions'
export const BANK_ACCOUNT_STATS_DOCTYPE = 'io.cozy.bank.accounts.stats'
export const CONTACT_DOCTYPE = 'io.cozy.contacts'
export const RECURRENCE_DOCTYPE = 'io.cozy.bank.recurrence'
export const IDENTITIES_DOCTYPE = 'io.cozy.identities'
export const TAGS_DOCTYPE = 'io.cozy.tags'

export const offlineDoctypes = [
  ACCOUNT_DOCTYPE,
  GROUP_DOCTYPE,
  TRANSACTION_DOCTYPE,
  SETTINGS_DOCTYPE,
  BILLS_DOCTYPE,
  CONTACT_DOCTYPE,
  TAGS_DOCTYPE
]

class HasManyBills extends HasManyInPlace {
  get data() {
    return this.raw
      ? this.raw.map(doctypeId => {
          const [doctype, id] = doctypeId.split(':')
          return this.get(doctype, id)
        })
      : []
  }

  /**
   * Query is redefined since the ids are prepended with their doctype.
   * Also does not refetch if documents have already been fetched once.
   */
  static query(doc, client, assoc) {
    const included = doc[assoc.name]
    if (!included || !included.length) {
      return null
    }

    const ids = included.indexOf(':')
      ? included.map(x => x.split(':')[1])
      : included

    const docs = fromPairs(
      ids.map(id => {
        return [id, client.getDocumentFromState(assoc.doctype, id)]
      })
    )

    const missingIds = Object.keys(docs).filter(id => !docs[id])
    if (!missingIds || !missingIds.length) {
      return Object.values(docs)
    } else {
      return new QueryDefinition({ doctype: assoc.doctype, ids: missingIds })
    }
  }
}

export class HasManyReimbursements extends HasManyInPlace {
  get raw() {
    return this.target[this.name]
  }

  get data() {
    return (this.raw || []).map(reimbursement => {
      if (!reimbursement.billId) {
        return reimbursement
      }
      return {
        ...reimbursement,
        bill: this.get('io.cozy.bills', reimbursement.billId.split(':')[1])
      }
    })
  }

  static query(doc, client, assoc) {
    const included = doc[assoc.name]
    if (!included || !included.length) {
      return null
    }
    const missingIds = included
      .map(doc => doc.billId && doc.billId.split(':')[1])
      .filter(Boolean)

    return new QueryDefinition({ doctype: assoc.doctype, ids: missingIds })
  }
}

// Add transactions relationships on tags
export class HasManyTags extends HasMany {
  async add(docsArg) {
    const docs = Array.isArray(docsArg) ? docsArg : [docsArg]
    const targets = Array.isArray(this.target) ? this.target : [this.target]

    for (const tag of docs) {
      await tag.transactions?.add(targets)
    }

    return super.add(docs)
  }

  async remove(docsArg) {
    const docs = Array.isArray(docsArg) ? docsArg : [docsArg]
    const targets = Array.isArray(this.target) ? this.target : [this.target]

    for (const tag of docs) {
      await tag.transactions?.remove(targets)
    }

    return super.remove(docs)
  }
}

// Add tags relationships on transactions
export class HasManyTransactions extends HasMany {
  async add(docsArg) {
    const docs = Array.isArray(docsArg) ? docsArg : [docsArg]
    const targets = Array.isArray(this.target) ? this.target : [this.target]

    for (const transaction of docs) {
      await transaction.tags?.add(targets)
    }

    return super.add(docs)
  }

  async remove(docsArg) {
    const docs = Array.isArray(docsArg) ? docsArg : [docsArg]
    const targets = Array.isArray(this.target) ? this.target : [this.target]

    for (const transaction of docs) {
      await transaction.tags?.remove(targets)
    }

    return super.remove(docs)
  }
}

const addTag = flag('banks.tags.enabled')
  ? {
      tags: {
        type: HasManyTags,
        doctype: TAGS_DOCTYPE
      }
    }
  : {}
export const schema = {
  transactions: {
    doctype: TRANSACTION_DOCTYPE,
    attributes: {},
    relationships: {
      account: {
        type: 'belongs-to-in-place',
        doctype: ACCOUNT_DOCTYPE
      },
      recurrence: {
        type: 'has-one',
        doctype: RECURRENCE_DOCTYPE
      },
      bills: {
        type: HasManyBills,
        doctype: BILLS_DOCTYPE
      },
      reimbursements: {
        type: HasManyReimbursements,
        doctype: BILLS_DOCTYPE
      },
      ...addTag
    }
  },
  bills: {
    doctype: BILLS_DOCTYPE
  },
  settings: {
    doctype: SETTINGS_DOCTYPE,
    attributes: {},
    relationships: {}
  },
  bankAccounts: {
    doctype: ACCOUNT_DOCTYPE,
    attributes: {},
    relationships: {
      checkingsAccount: {
        type: 'has-one',
        doctype: ACCOUNT_DOCTYPE
      },
      owners: {
        type: 'has-many',
        doctype: CONTACT_DOCTYPE
      },
      connection: {
        type: 'has-one',
        doctype: COZY_ACCOUNT_DOCTYPE
      }
    }
  },
  groups: {
    doctype: GROUP_DOCTYPE,
    attributes: {},
    relationships: {
      accounts: {
        type: HasManyInPlace,
        doctype: ACCOUNT_DOCTYPE
      }
    }
  },
  triggers: {
    doctype: TRIGGER_DOCTYPE,
    attributes: {},
    relationships: {}
  },
  apps: {
    doctype: APP_DOCTYPE,
    attributes: {},
    relationships: {}
  },
  konnectors: {
    doctype: KONNECTOR_DOCTYPE,
    attributes: {},
    relationships: {}
  },
  stats: {
    doctype: BANK_ACCOUNT_STATS_DOCTYPE,
    attributes: {},
    relationships: {
      account: {
        type: 'has-one',
        doctype: ACCOUNT_DOCTYPE
      }
    }
  },
  tags: {
    doctype: TAGS_DOCTYPE,
    relationships: {
      transactions: {
        type: HasManyTransactions,
        doctype: TRANSACTION_DOCTYPE
      }
    }
  }
}

export const older30s = CozyClient.fetchPolicies.olderThan(30 * 1000)
export const neverReload = CozyClient.fetchPolicies.olderThan(100000 * 1000)

export const accountsConn = {
  query: () => Q(ACCOUNT_DOCTYPE).include(['owners', 'connection']),
  as: 'accounts',
  fetchPolicy: older30s
}

export const groupsConn = {
  query: () => Q(GROUP_DOCTYPE),
  as: 'groups',
  fetchPolicy: neverReload
}

export const cronKonnectorTriggersConn = {
  query: () =>
    Q(TRIGGER_DOCTYPE).where({
      worker: 'konnector',
      type: '@cron'
    }),
  as: 'io.cozy.triggers/konnector_cron',
  fetchPolicy: neverReload,
  hydrated: false,
  singleDocData: false
}

export const transactionsConn = {
  query: () =>
    Q(TRANSACTION_DOCTYPE)
      .UNSAFE_noLimit()
      .include(['bills', 'account', 'reimbursements', 'recurrence', 'tags']),
  as: 'transactions',
  fetchPolicy: older30s
}

export const makeBalanceTransactionsConn = () => {
  const fromDate = format(subYears(new Date(), 1), 'YYYY-MM-DD')
  return {
    as: 'home/transactions',
    fetchPolicy: neverReload,
    query: () =>
      Q(TRANSACTION_DOCTYPE)
        .limitBy(1000)
        .where({
          date: {
            $gt: fromDate
          }
        })
        .sortBy([{ date: 'desc' }])
        .indexFields(['date'])
        .select([
          '_id',
          'date',
          'realisationDate',
          'amount',
          'account',
          'currency',
          'manualCategoryId',
          'cozyCategoryId',
          'cozyCategoryProba',
          'localCategoryId',
          'localCategoryProba',
          'automaticCategoryId',
          'label',
          'reimbursementStatus',
          'reimbursements'
        ])
  }
}

export const appsConn = {
  query: () => Q(APP_DOCTYPE),
  as: 'apps',
  fetchPolicy: neverReload
}

export const billsConn = {
  query: () => Q(BILLS_DOCTYPE),
  as: 'bills',
  fetchPolicy: older30s
}

export const settingsConn = {
  query: () => Q(SETTINGS_DOCTYPE),
  as: 'settings',
  fetchPolicy: older30s
}

export const recipientsConn = {
  query: () => Q(RECIPIENT_DOCTYPE),
  as: 'recipients',
  fetchPolicy: older30s
}

export const recurrenceConn = {
  query: () =>
    Q(RECURRENCE_DOCTYPE)
      .where({ _id: { $gt: null }, latestDate: { $gt: null } })
      .sortBy([{ _id: 'desc' }, { latestDate: 'desc' }]),
  as: 'recurrence',
  fetchPolicy: older30s
}

export const myselfConn = {
  query: () => Q('io.cozy.contacts').where({ me: true }),
  as: 'myself',
  fetchPolicy: older30s
}

export const konnectorConn = {
  query: slug => Q(KONNECTOR_DOCTYPE).getById(`${KONNECTOR_DOCTYPE}/${slug}`),
  as: 'konnector'
}

export const tagsConn = {
  query: () => Q(TAGS_DOCTYPE).include(['transactions']),
  as: `${TAGS_DOCTYPE}/withTransactions`,
  fetchPolicy: older30s
}

export const buildTagsQueryWithTransactionsByIds = ids => {
  return {
    definition: Q(TAGS_DOCTYPE).getByIds(ids).include(['transactions']),
    options: {
      as: `${TAGS_DOCTYPE}/${JSON.stringify(ids)}/withTransactions`,
      fetchPolicy: older30s
    }
  }
}

export const buildTransactionsWithTagsQueryByIds = ids => {
  return {
    definition: Q(TRANSACTION_DOCTYPE).getByIds(ids).include(['tags']),
    options: {
      as: `${TRANSACTION_DOCTYPE}/${JSON.stringify(ids)}/withTags`,
      fetchPolicy: older30s
    }
  }
}
