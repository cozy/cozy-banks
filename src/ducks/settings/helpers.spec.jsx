import React from 'react'
import { mount } from 'enzyme'
import fixtures from 'test/fixtures/unit-tests.json'
import { ACCOUNT_DOCTYPE, GROUP_DOCTYPE, SETTINGS_DOCTYPE } from 'doctypes'
import AppLike from 'test/AppLike'
import CozyClient, { createMockClient } from 'cozy-client'

import {
  fetchSettings,
  withAccountOrGroupLabeller,
  getWarningLimitsPerAccount,
  reverseIndex,
  disableOutdatedNotifications,
  downloadFile
} from './helpers'

describe('defaulted settings', () => {
  it('should return defaulted settings', async () => {
    const fakeClient = {
      find: () => {},
      query: () => {
        return Promise.resolve({
          data: [
            {
              pin: '1234'
            },
            {
              notifications: {
                balanceLower: {
                  value: 600,
                  enabled: false
                }
              }
            }
          ]
        })
      }
    }
    const settings = await fetchSettings(fakeClient)
    expect(settings).toMatchSnapshot()
  })
})

describe('withAccountOrGroupLabeller', () => {
  const setup = ({ accountOrGroup }) => {
    const client = createMockClient({
      queries: {
        groups: {
          doctype: GROUP_DOCTYPE,
          data: fixtures[GROUP_DOCTYPE]
        },
        accounts: {
          doctype: ACCOUNT_DOCTYPE,
          data: fixtures[ACCOUNT_DOCTYPE]
        }
      }
    })

    const DumbComponent = ({ accountOrGroup, getAccountOrGroupLabel }) => {
      return <div>{getAccountOrGroupLabel(accountOrGroup)}</div>
    }

    const Component = withAccountOrGroupLabeller('getAccountOrGroupLabel')(
      DumbComponent
    )

    const root = mount(
      <AppLike client={client} store={client.store}>
        <Component accountOrGroup={accountOrGroup} />
      </AppLike>
    )
    return { root }
  }

  it('should correctly name an account', () => {
    const { root } = setup({
      accountOrGroup: {
        _id: 'compteisa1',
        _type: ACCOUNT_DOCTYPE
      }
    })
    expect(root.text()).toBe('Compte courant Isabelle')
  })

  it('should correctly name a group', () => {
    const ENLARGED_FAMILY_GROUP_ID = 'familleelargie'
    const { root } = setup({
      accountOrGroup: {
        _id: ENLARGED_FAMILY_GROUP_ID,
        _type: GROUP_DOCTYPE
      }
    })
    expect(root.text()).toBe('Famille Ã©largie')
  })

  it('should correctly name an autogroup', () => {
    const { root } = setup({
      accountOrGroup: {
        _id: 'autogroup1',
        _type: GROUP_DOCTYPE
      }
    })
    expect(root.text()).toBe('Checking accounts')
  })
})

describe('reverseIndex', () => {
  it('should work', () => {
    const homer = { keys: [1, 2, 3], name: 'homer' }
    const marge = { keys: [4, 2, 6], name: 'marge' }
    expect(reverseIndex([homer, marge], x => x.keys)).toEqual({
      1: [homer],
      2: [homer, marge],
      3: [homer],
      4: [marge],
      6: [marge]
    })
  })
})

describe('getWarningLimitsPerAccount', () => {
  it('should return warningLimitsPerAccount from rules, groups and accounts', () => {
    const groups = fixtures[GROUP_DOCTYPE]
    const accounts = fixtures[ACCOUNT_DOCTYPE]
    const rules = [
      {
        value: 75,
        accountOrGroup: {
          _type: GROUP_DOCTYPE,
          _id: 'familleelargie'
        },
        enabled: true
      },
      {
        value: 50,
        accountOrGroup: {
          _type: GROUP_DOCTYPE,
          _id: 'isabelle'
        },
        enabled: true
      },
      {
        value: 100,
        accountOrGroup: {
          _type: ACCOUNT_DOCTYPE,
          _id: 'comptelou1'
        },
        enabled: true
      },
      {
        value: 10,
        enabled: true
      },
      {
        value: 15
      }
    ]
    const warningLimits = getWarningLimitsPerAccount(rules, accounts, groups)
    expect(warningLimits['comptelou1']).toBe(100)
    expect(warningLimits['compteisa3']).toBe(50)
    expect(warningLimits['compteisa1']).toBe(75)
    expect(warningLimits['comptegene1']).toBe(75)
    // isa4 is not in the isabelle group
    expect(warningLimits['compteisa4']).toBe(10)
  })
})

describe('remove account from notifications', () => {
  it('should remove account from notifications when an account is deleted', async () => {
    jest.spyOn(console, 'info').mockImplementation(() => {})
    const client = new CozyClient()
    const oldSettings = {
      _rev: 'rev-1',
      ...fixtures[SETTINGS_DOCTYPE][0]
    }
    client.queryAll = jest
      .fn()
      .mockResolvedValueOnce({
        data: [{ _id: 'compteisa1', _type: ACCOUNT_DOCTYPE, _deleted: true }]
      })
      .mockResolvedValueOnce({ data: [] })
      .mockResolvedValue({ data: [oldSettings] })

    client.save = jest.fn()
    const disabledNotifs = await disableOutdatedNotifications(client)

    expect(client.save).toHaveBeenCalledWith(
      expect.objectContaining({
        notifications: expect.objectContaining({
          balanceLower: [
            expect.objectContaining({
              id: 0,
              enabled: false,
              accountOrGroup: null,
              value: 100
            })
          ]
        })
      })
    )

    expect(disabledNotifs).toEqual([
      {
        accountOrGroup: {
          _id: 'compteisa1',
          _type: 'io.cozy.bank.accounts'
        },
        enabled: true,
        id: 0,
        value: 100
      },
      {
        accountOrGroup: {
          _id: 'compteisa1',
          _type: 'io.cozy.bank.accounts'
        },
        enabled: true,
        id: 0,
        value: 1000
      }
    ])
  })
})

describe('downloadFile', () => {
  const setup = ({
    mockQueued = jest.fn(),
    mockCreate = jest.fn(),
    mockDownload = jest.fn()
  } = {}) => {
    const client = {
      collection: jest.fn(() => ({
        queued: mockQueued,
        create: mockCreate,
        download: mockDownload
      }))
    }
    return client
  }
  it('should early return if file is falsy', async () => {
    const mockDownload = jest.fn(() => 'download_link')
    setup({ mockDownload })

    await downloadFile(undefined)

    expect(mockDownload).toBeCalledTimes(0)
  })

  it('should call "download" with correct arguments ', async () => {
    const mockFile = { id: '00', name: 'mockFileName' }
    const mockDownload = jest.fn(() => 'download_link')
    const client = setup({ mockDownload })

    await downloadFile(client, mockFile)

    expect(mockDownload).toBeCalledTimes(1)
    expect(mockDownload).toBeCalledWith(mockFile)
  })
})
