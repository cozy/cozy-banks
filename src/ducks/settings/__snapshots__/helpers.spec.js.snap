// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaulted settings should return defaulted settings 1`] = `
Object {
  "_type": "io.cozy.bank.settings",
  "billsMatching": Object {
    "billsLastSeq": "0",
    "transactionsLastSeq": "0",
  },
  "categorization": Object {
    "lastSeq": 0,
  },
  "community": Object {
    "autoCategorization": Object {
      "enabled": false,
    },
    "localModelOverride": Object {
      "enabled": false,
    },
  },
  "notifications": Object {
    "balanceLower": Object {
      "enabled": false,
      "value": 600,
    },
    "healthBillLinked": Object {
      "enabled": true,
    },
    "hebdo": Object {
      "enabled": false,
    },
    "lastSeq": 0,
    "mensuel": Object {
      "enabled": false,
    },
    "salaire": Object {
      "enabled": false,
    },
    "transactionGreater": Object {
      "enabled": true,
      "value": 600,
    },
  },
  "showIncomeCategory": true,
}
`;
